### PR TITLE
Expire share after end of day

### PR DESCRIPTION
**Pull request by phil-davis**
## Description
With the current code in #36573 the share expires at the very start of the day specified. e.g. set the expiration date to "2019-12-20" and when you get to the office on 2019-12-20 the share is already expired and gone.

You can set the expiry date to today (and set the default expiry days to 0, and enforce that - any of those). If you then create a share that expires today, then the share gets created but effectively expires at the beginning of today. As soon as the share receiver goes to access the share, the backend code detects that the share is already expired and it deletes the share. That is all a bit silly.

This PR changes the meaning of the expiry so that "expire date 2019-12-20" means that the share expires at the end of "2019-12-20" - i.e. the moment that the date ticks over to "2019-12-21" then the share will get deleted.

That allows shares created with expire date "today" to work in a reasonable fashion.

And also maybe users will be happier that a share expiring on some future date will still work until the end of that day.

Edit: see https://github.com/owncloud/product/blob/master/modules/ROOT/examples/oc10/collaborators/expiration_date_for_collaborators.feature#L35 which specifies `WHEN the specified day has just ended` - that is the behaviour that this PR is implementing.

## Related Issue
- #36569 

## Motivation and Context
Adjust weird behaviour.

## How Has This Been Tested?
Local test runs of the acceptance test and unit tests that were changed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
As well as code review, we need to decide if this is the required behaviour.

**Comment by phil-davis** [comment]
Note: JS tests are failing because they are currently failing in the underlying PR

File: lib/private/Share20/Manager.php [hunk 1]
```diff
@@ -1361,15 +1369,16 @@ public function getAllSharedWith($userId, $shareTypes, $node = null) {
 		// Aggregate all required $shareTypes by mapping provider to supported shareTypes
 		$providerIdMap = $this->shareTypeToProviderMap($shareTypes);
 
-		$today = new \DateTime();
+		$today = new \DateTime("today");
```
Comment (jvillafanez) [inline lib/private/Share20/Manager.php]: Not needed

Comment (phil-davis) [reply lib/private/Share20/Manager.php]: removed, thanks for spotting

**Comment by jvillafanez** [comment]
I think the problem is that we're setting the expiration date as `2020-1-12 00:00:00`, and we're probably overwriting the time (sending `2020-1-10 11:11:11` would be converted to `2020-1-10 00:00:00`).

I'd prefer to set the full datetime properly in the DB or set `23:59:59` as time if none is provided (`2020-1-11` would be converted to `2020-1-11 23:59:59`). This is probably more consistent with the current code, but it's probably more work.

In any case, I don't have a strong opinion on this if we decide just to focus on the specific date.

**Comment by phil-davis** [comment]
Yes, I am not sure if we are setting the time in the database to 00:00:00 or to the current time.

In the acceptance tests we create a share, e.g. at 2019-12-16:09:15:30:00, with expiry date sent as "2019-12-16". The database might record 2019-12-16:09:15:30:00 or 2019-12-16:00:00:00:00 as the expiry date.

Then 1 second later at 2019-12-16:09:15:30:01 we check the receiver's view of the share. Either way the share is expired.

**Comment by jvillafanez** [comment]
I'll ping @micbar to decide what to do. From my point of view, this is mergeable.

I'd recommend to merge this PR with the whole feature, and not block the release. We can open an enhancement to support specifying the time so you could set `2020-1-3 12:30:00` as expiration, to be prioritized after the release.

**Comment by phil-davis** [comment]
CI passed (apart from the known JS test problem)
The underlying branch was rebased and the JS test problem was fixed. I squashed the 4 commits that were here into 1 commit and rebased it on top of the feature branch, so this should all be passing now.

**Comment by phil-davis** [comment]
This has been cherry-picked into PR #36573

**Pull request closed without merging**